### DISCUSSION TITLE
TressFX or other hair system

### DISCUSSION THREAD

**Discussion by Himhoji** [godotengine/godot-proposals#8103] [3D]
When Tomb Raider 2013 came out, it had this shiny new tech named TressFX.
Since then, I only saw it in some tech demos and it only got used in three games that I know of: 
- Tomb Raider 2013
- Deus Ex: Mankind Divided
- 逆水寒（Justice)

The Tomb Raider 2013 sequels had their own version called PureHair, which was built from TressFX.

I always asked myself why this tech wasn't in use more often.
I think it would be a huge step in graphics for Godot, if it were to be implemented.
I certainly miss something like this in my projects.

But I have little knowledge about the possibilities of implementing TressFX into Godot and how sensical that would be.
What are your thoughts about this?
I would love to have a hair system in the engine.

**Comment by AThousandShips** [comment]
This is the area for a plugin, not core, far too specific

**Comment by Himhoji** [comment]
Well, I don't know about others, but I would use it all the time.
And many creatures I saw in games made with Godot would definitly benefit from some fur.
People love themselves some fur.

But on a serious note. I really think with how many games are made with creatures or grass (or bacteria), this wouldn't be a feature used seldom.

**Reply by AThousandShips** [reply]
That's not what I mean with "specific" though, I mean that it's something that's hard to make fit everyone, further most games won't use hair or grass like this

This being part of the engine would mean finding or making a system that fits most styles people want, and most performance requirements, that's incredibly hard

**Reply by Himhoji** [reply]
Oh, ok I misunderstood.
What exactly do you mean with "...won't use hair or grass 'like this' "?

Is there any other hair system solution of which inspiration could be drawn from?

**Reply by AThousandShips** [reply]
Doesn't need physical hair or grass

**Reply by Himhoji** [reply]
Sorry, I don't understand.

**Reply by AThousandShips** [reply]
Most games made with Godot doesn't need hair that works like this, or grass, just using static models, not physics, no hair system at all

Just using a mesh for hair or grass, just solid

**Reply by Himhoji** [reply]
I see, thank you for specifying.
Personally I think that is not very future oriented.
I hope we will get a hair system some time in the near future. Be it as extension or some other way.

**Reply by AThousandShips** [reply]
Future how? How is not using hair in *your own game* not future oriented?

**Reply by Himhoji** [reply]
I'm sorry, now I don't understand what you are asking me.

**Reply by AThousandShips** [reply]
> Personally I think that is not very future oriented.

I said that many won't use this feature, and you said that's not future oriented, so I'm confused

**Reply by Himhoji** [reply]
Ah, we simply have a disagreement then. Because I think that a good chunk of people would use this feature.
That is why I said, that it would't be future oriented to not implement such a feature, from my perspective.

And since you don't see the point of this feature being implemented at the moment, you don't think it would be a huge loss to not implement such a feature.

Is that right?

**Reply by AThousandShips** [reply]
You are putting words in my mouth, read my comments as to why I don't think this should be added as a feature *to the engine* but should be as a plugin, I have not said I don't see the point of it, I very much do, please read my comments again and see what I am actually saying :)

But I will repeat them:
* Hair and grass physics is something that's complicated to get right
* It varies a lot between projects, there's no "one size fits all" solution, there's also varying levels of complexity and cost
* Not every 3D game would need this, many indie games wouldn't
* This can be served much better by a plugin, or many plugins for each type of hair or grass, for various levels of performance

**Reply by Himhoji** [reply]
By grace, I said this as reference to implementing it in or to the core. I simply asked if that was what you meant.
I wasn't putting any words in your mouth.
This is a very strange conversation.

But I appreciate your contributions and thank you for the list. It is very helpful! :D

**Comment by theraot** [comment]
I don't know the details of TressFX.

Yet, I would like to see some kind of hair/fur system in Godot. And yes, in core, as I believe it would be used often if it was just there. However, it sounds like it needs the right know how to implement and maintain. Which makes me think it is unlikely to happen anytime soon.

Meanwhile, I suppose that if somebody out there - with the knowledge to implement this - is willing to put the time and effort, this feature could begin its life as a GDExtension plugin (in particular if we are talking about integrating a third party solution such as TressFX instead of something tailored for Godot).

**Reply by Calinou** [reply]
> Yet, I would like to see some kind of hair/fur system in Godot.

Good old shell fur can get you pretty far: https://github.com/Arnklit/ShellFurGodot

**Reply by Himhoji** [reply]
Awesome! I didn't know there was shell fur for Godot. Good to know!

**Comment by chiefwhosm** [comment]
I was looking at information on how hair is done in Godot and seeing that it's only hair cards or physics bones, found this proposal in my searches.

In terms of TressFX...

The O3DE engine (DirectX 12 and Vulkan, Forward+ renderering) has their implementation under the MIT and Apache 2.0 licences which may give someone with the know-how enough information to port TressFX into Godot:
https://github.com/o3de/o3de/tree/development/Gems/AtomTressFX

Also there was partial port of TressFX to OpenGL (again MIT licence), which possibly would be useful for the mobile and compatibility renderer?
https://github.com/Scthe/TressFX-OpenGL

And of course, there's still the original MIT licenced source code from AMD:
https://github.com/GPUOpen-Effects/TressFX

One issue is generating hair in a 3rd party application, or rather exporting it to the engine. Most export scripts are for older versions of Maya and Blender so would likely need updating to support new updates, this is the most recent feature-complete exporter I know of, but it's for Blender v2.79:
https://github.com/kostenickj/TressFXExporter

As an alternative to TressFX, I recently found this video on YouTube which uses mesh shaders for hair generation, and while if I'm right, this is something Godot has yet to implement? this might be a viable alternative, albeit meaning a solution built from scratch and new tools made to allow hair creation either in or out of Godot itself:
https://www.youtube.com/watch?v=LwgOfPbDYRU
(I do notice the hair sometimes stabs through the mesh  from one area to another though).

Finally, there's the old 2008 nVidia Nalu Mermaid hair, it looks similar to how TressFX works, but figured I'd add this to the available suggestions:
https://www.youtube.com/watch?v=RChiNlG2_oc
(https://developer.nvidia.com/gpugems/gpugems2/part-iii-high-quality-rendering/chapter-23-hair-animation-and-rendering-nalu-demo)

**Comment by Jesusemora** [comment]
You can achieve hair physics with SpringBoneModifier and bones. They have limitations, but can range from very simple (one chain of bones for all the hair) to more complex (one per strand).
Personaly, hair physics sound like some AAA feature that needs a powerful PC to run and a lot of work to implement in your game. If this can simplify the workflow, then it would be better than springbones.

**Reply by chiefwhosm** [reply]
In regards to needing a powerful PC, I have to admit I was impressed by TressFX since I was able to play the 2013 TombRaider game (1080p), with TressFX enabled on my old computer from 2011:
2.3Ghz dual core Intel Core i5, 8gb ram and a 256megabyte VRAM Radeon HD 6630M

**Comment by ismailivanov** [comment]
I ported AMD TressFX to Godot 4 [TressFX-Godot](https://github.com/ismailivanov/TressFX-Godot)

**Reply by Himhoji** [reply]
Thank you very much; this is awesome!
And quick with the updates too :0
Can't wait to use this.
